### ISSUE TITLE
Request the SPOA support the the V3 modsecurity

### ISSUE THREAD

**Issue by leibaogit** [haproxy/haproxy#1021]
## What should haproxy do differently? Which functionality do you think we should add?

SPOA support for the [modsecurity V3](https://github.com/SpiderLabs/ModSecurity)

## What are you trying to do?

I'm trying to integrate the HAProxy community version with the modsecurity v3 with the SPOA, but now I can only found the SPOA support for the [V2 modsecurity](https://github.com/haproxy/haproxy/tree/master/contrib/modsecurity).

## Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv
HA-Proxy version 1.8.23 2019/11/25
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-format-truncation -Wno-null-dereference -Wno-unused-label -Wno-stringop-overflow
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
Running on OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.4
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.42 2018-03-20
Running on PCRE version : 8.42 2018-03-20
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace

# uname -a
Linux vsi-haproxy 4.18.0-193.14.2.el8_2.x86_64 #1 SMP Sun Jul 26 03:54:29 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by wtarreau** [comment]
And what about trying to implement it yourself ? The V2 SPOA agent was provided entirely for illustration purposes and serves this purpose pretty well. It's self contained and shows pretty well how to interface with a library. For modsecv3 they've split the code into connectors and they're using the ModsecurityNginx connector as an example for the same reasons. Thus I guess it should not be that hard to perform sort of a merge of the two and get a ModsecuritySPOA connector. Maybe it could even interest the modsecurity project to maintain it and make it follow the internal API changes over time to keep it as a useful example, I don't know.

Just my two cents.

**Comment by leibaogit** [comment]
@wtarreau yeah, I want to implement it, but, I'm new one, and may need guidance or help. 

Now I have two options if implement it:
1. Still using the SPOA, just as you said, using the current V2 SPOA agent for reference. 
2. You know, the V3 modsecurity has split the core mode to libmodsecurity, and using the connector to connect to different web servers. So I'm wondering whether we can implement a haproxyConnector to connect the libmodsecurity to haproxy.

Seems the option 2 is much better, it's a static function call, no protocols engaged. So what's your opinion?

**Comment by wtarreau** [comment]
No, option 2 is the worst one for a product like modsec. The reason SPOE was created is for such long processing which is totally incompatible with event-driven, low-latency systems. The longest request I've seen on modsec took 66 **minutes** to analyse (yes, more than one hour)! And requests taking hundreds of milliseconds to seconds are not uncommon at all there, while even a millisecond is already considered as infinitely long for request processing. These ones totally destroy a low-latency product like haproxy because nothing else can happen in parallel anymore, all traffic suffers a total freeze, even if not required to pass through the WAF. Timeouts accumulate, connections are dropped on the frontend, etc. This simply is not an option at all, and there's no way to address it because the model is fundamentally wrong.

With SPOE, only the traffic subject to the analyze is submitted outside, in a scalable way (you can put as many instances of your WAF as you need), and there is zero effect on the rest of the traffic passing through the proxy. This is absolutely critical for such type of high-latency processing. Other good examples of usage of such a thing would be image resizing or recompression for example, but also a wide range of complex authentication mechanisms. We've one thought that DNS resolution could have been a good candidate, but DNS is not too complex to implement in an event-driven way and could be implemented natively.

Thus no, please really focus on option 1. It's by far much better and the only one that will be acceptable. Plus it comes with other benefits like being version independent, and not requiring to restart the proxy each time you change your WAF settings or rules, and conversely.

There's no problem with being new here. Take your time to observe how things work and to figure how to best assemble the elements you have. You'll certainly make some mistakes initially, this is expected and not a problem at all, so don't worry a single second about this :-)

**Comment by leibaogit** [comment]
@wtarreau Thanks for your sophisticated explanation, and it saved me out of the struggle of the two options :)

Then let's talk about the SPOE solution. V3 libmodsecurity can use different connectors to connect with different web servers (Nginx, IIS, Apache), then want to know whether the SPOA work with web server or libmodsecurity? If works with web server, then does it need different types SPOA to adept different web servers?

**Comment by wtarreau** [comment]
The SPOA is the agent (the one running the heavy work), SPOE is the client engine (the HTTP side, haproxy here). SPOP is the protocol used between the two, even though by abuse of language we often call it SPOE as well (you'll find all these details in doc/SPOE.txt).

So the SPOA is technically a daemon, a listening TCP server accepting connections from haproxy, over which some requests for content analysis will be sent. This daemon will then possibly reconstruct a full HTTP-like request if required by modsec, or more likely a very simplified one, and submit that to libmodsecurity, will wait for the result, then will send the verdict back to haproxy.

As such, the SPOA doesn't depend on any server and provides its own listening service. It could, however, be used by other servers or proxies that implement the SPOP protocol. Its interfacing with libmodsecurity will be implemented exactly as if it were itself a web server submitting the requests it receives to the WAF, except that its frontend protocol is not HTTP but SPOP. A pretty similar architecture with such a sidecar analyser running out of the proxies is common with the ICAP protocol by the way.

Let me know if something is not clear.

**Comment by leibaogit** [comment]
@wtarreau sorry for a late response I was stuck in other things for few days :-(.... Thanks again for the explanation, and I've asked the modsecurity in this [issue](https://github.com/SpiderLabs/ModSecurity/issues/2491), and got an answer that there is a project doing a similar thing, they use the [golang ModSecurity library bindings](https://github.com/rikatz/go-modsecurity) to expand their haproxy, and in their case, they are building this [ingress-security-agent](https://github.com/rikatz/ingress-security-agent/).

So may I can start from these, and are you ware of the project, any suggestions if leverage them to build standalone SPOE for v3 modsecurity?
